core/src/main/java/org/apache/calcite/rel/core/Join.java

```diff
@@ -52,11 +55,12 @@
  * The set of output rows is a subset of the cartesian product of the two
  * inputs; precisely which subset depends on the join condition.
  */
-public abstract class Join extends BiRel {
+public abstract class Join extends BiRel implements RelWithHint {
```

Comment:
Will hint be part of relnode digest?

Reply:
For current implementation it is not, so there may be some problem when there are two same digest nodes that just differ with hints during planner planning, i know it is a problem, i'm just hesitant to add hints to digest because it would be very redundant and plan would looks too complex.(Not every hints on the node would meaningful, say, it's just a hint and not forced attribute).

I'm considering if i can give an internal implementation for digest + hints node reuse without modify the node digests.